File: xdm-core/src/main/java/xdm/core/downloaders/DownloadProgressData.java
package xdm.core.downloaders;

import java.util.EnumSet;
import java.util.LinkedList;
import java.util.List;
import lombok.Getter;
import lombok.Setter;
import xdm.core.util.FormatUtilities;

public class DownloadProgressData {
  private final AbstractDownloader downloader;
  @Setter private long totalDownloadedBytes;
  private long lastDownloadedBytes;
  private final LinkedList<Double> speedHistory;
  private double speedSum;
  private long downloadedBytesSinceStartOrResume,
      ticksAtDownloadStartOrResume,
      lastProgressUpdatedAt,
      lastChunkStatsUpdated;
  @Getter private int progress;
  @Getter private float downloadSpeed;
  @Getter private long eta;
  @Getter private SegmentDetails segmentDetails;
  @Setter private int lastProgress;
  private final EnumSet<DownloaderType> adaptiveDownloaderTypes =
      EnumSet.of(DownloaderType.Hls, DownloaderType.Dash, DownloaderType.EncryptedHls);

  public DownloadProgressData(AbstractDownloader downloader, long totalDownloadedBytes) {
    this.downloader = downloader;
    this.totalDownloadedBytes = totalDownloadedBytes;
    this.ticksAtDownloadStartOrResume = System.currentTimeMillis();
    this.downloadedBytesSinceStartOrResume = 0;
    this.speedHistory = new LinkedList<>();
  }

  public synchronized void updateDownloadInfo(long downloadedBytes, List<Chunk> chunks) {
    long ticks = System.currentTimeMillis();
    this.totalDownloadedBytes += downloadedBytes;
    this.downloadedBytesSinceStartOrResume += downloadedBytes;
    long ticksElapsed = ticks - this.ticksAtDownloadStartOrResume;
    if (ticks - this.lastProgressUpdatedAt > 500 && ticksElapsed > 0) {
      double instantSpeed =
          ((this.totalDownloadedBytes - this.lastDownloadedBytes) * 1000)
              / (double) (ticks - this.lastProgressUpdatedAt);
      if (this.speedHistory.size() > 5) {
        this.speedSum -= this.speedHistory.removeFirst();
      }
      this.speedHistory.add(instantSpeed);
      speedSum += instantSpeed;
      double instanceSpeedBounded = speedSum / this.speedHistory.size();
      double avgSpeed = (this.downloadedBytesSinceStartOrResume * 1000.0) / ticksElapsed;
      this.lastProgressUpdatedAt = ticks;
      this.lastDownloadedBytes = this.totalDownloadedBytes;
      this.progress = downloader.getProgress().get();
      this.downloadSpeed = (float) instanceSpeedBounded;
      if (adaptiveDownloaderTypes.contains(downloader.getDownloaderType())) {
        int prgDiff = progress - lastProgress;
        lastProgress = progress;
        if (prgDiff > 0) {
          this.eta = (ticksElapsed * (100 - progress) / 1000 * prgDiff);
        }
      } else {
        this.eta =
            FormatUtilities.getEtaAsSec(
                (double) downloader.getSize() - totalDownloadedBytes, (float) avgSpeed);
      }
    }
    if (ticks - this.lastChunkStatsUpdated > 1000 && ticksElapsed > 0) {
      this.updateChunkProgressData(chunks);
      this.lastChunkStatsUpdated = ticks;
    }
  }

  private void updateChunkProgressData(List<Chunk> chunks) {
    if (segmentDetails == null) {
      segmentDetails = new SegmentDetails();
    }
    if (this.adaptiveDownloaderTypes.contains(this.downloader.getDownloaderType())) {
      if (segmentDetails.getCapacity() < downloader.getMaxCount()) {
        segmentDetails.extend(downloader.getMaxCount());
      }
      int prg = downloader.getProgress().get();
      for (int i = 0; i < downloader.getMaxCount(); i++) {
        SegmentInfo info = segmentDetails.getChunkUpdates().get(i);
        info.setDownloaded(prg);
        info.setStart(i * 100L);
        info.setLength(100);
      }
    } else {
      if (chunks == null) {
        return;
      }
      if (segmentDetails.getCapacity() < chunks.size()) {
        segmentDetails.extend(chunks.size() - segmentDetails.getCapacity());
      }
      segmentDetails.setChunkCount(chunks.size());
      for (int i = 0; i < chunks.size(); i++) {
        Chunk s = chunks.get(i);
        SegmentInfo info = segmentDetails.getChunkUpdates().get(i);
        info.setDownloaded(s.getDownloaded());
        info.setStart(s.getStartOffset());
        info.setLength(s.getLength());
      }
    }
  }
}

File: xdm-core/src/main/java/xdm/core/downloaders/DownloadProgressData.kt
//package xdm.core.downloaders
//
//import xdm.core.util.FormatUtilities
//import java.util.*
//
//class DownloadProgressData(
//    val downloader: AbstractDownloader, var totalDownloadedBytes: Long
//) {
//    var lastDownloadedBytes: Long = 0
//    val speedHistory = LinkedList<Double>()
//    var speedSum = 0.0
//    var downloadedBytesSinceStartOrResume: Long = 0
//    val ticksAtDownloadStartOrResume = System.currentTimeMillis()
//    var lastProgressUpdatedAt: Long = 0
//    var lastChunkStatsUpdated: Long = 0
//
//    var progress = 0
//
//    var downloadSpeed = 0f
//
//    var eta: Long = 0
//
//    var segmentDetails: SegmentDetails? = null
//
//    var lastProgress = 0
//    val adaptiveDownloaderTypes: EnumSet<DownloaderType> =
//        EnumSet.of(DownloaderType.Hls, DownloaderType.Dash, DownloaderType.EncryptedHls)
//
//    @Synchronized
//    fun updateDownloadInfo(downloadedBytes: Long, chunks: List<Chunk>?) {
//        val ticks = System.currentTimeMillis()
//        this.totalDownloadedBytes += downloadedBytes
//        this.downloadedBytesSinceStartOrResume += downloadedBytes
//        val ticksElapsed = ticks - this.ticksAtDownloadStartOrResume
//        if (ticks - this.lastProgressUpdatedAt > 500 && ticksElapsed > 0) {
//            val instantSpeed =
//                (((this.totalDownloadedBytes - this.lastDownloadedBytes) * 1000) / (ticks - this.lastProgressUpdatedAt).toDouble())
//            if (speedHistory.size > 5) {
//                this.speedSum -= speedHistory.removeFirst()
//            }
//            speedHistory.add(instantSpeed)
//            speedSum += instantSpeed
//            val instanceSpeedBounded = speedSum / speedHistory.size
//            val avgSpeed = (this.downloadedBytesSinceStartOrResume * 1000.0) / ticksElapsed
//            this.lastProgressUpdatedAt = ticks
//            this.lastDownloadedBytes = this.totalDownloadedBytes
//            this.progress = downloader.progress.get()
//            this.downloadSpeed = instanceSpeedBounded.toFloat()
//            if (adaptiveDownloaderTypes.contains(downloader.downloaderType)) {
//                val prgDiff = progress - lastProgress
//                lastProgress = progress
//                if (prgDiff > 0) {
//                    this.eta = (ticksElapsed * (100 - progress) / 1000 * prgDiff)
//                }
//            } else {
//                this.eta = FormatUtilities.getEtaAsSec(
//                    downloader.size.toDouble() - totalDownloadedBytes, avgSpeed.toFloat()
//                )
//            }
//        }
//        if (ticks - this.lastChunkStatsUpdated > 1000 && ticksElapsed > 0) {
//            this.updateChunkProgressData(chunks)
//            this.lastChunkStatsUpdated = ticks
//        }
//    }
//
//    fun updateChunkProgressData(chunks: List<Chunk>?) {
//        if (segmentDetails == null) {
//            segmentDetails = SegmentDetails()
//        }
//        if (adaptiveDownloaderTypes.contains(downloader.downloaderType)) {
//            if (segmentDetails!!.capacity < downloader.maxCount) {
//                segmentDetails!!.extend(downloader.maxCount)
//            }
//            val prg = downloader.progress.get()
//            for (i in 0..<downloader.maxCount) {
//                val info = segmentDetails!!.chunkUpdates[i]
//                info.downloaded = prg.toLong()
//                info.start = i * 100L
//                info.length = 100
//            }
//        } else {
//            if (chunks == null) {
//                return
//            }
//            if (segmentDetails!!.capacity < chunks.size) {
//                segmentDetails!!.extend(chunks.size - segmentDetails!!.capacity)
//            }
//            segmentDetails!!.chunkCount = chunks.size.toLong()
//            for (i in chunks.indices) {
//                val s = chunks[i]
//                val info = segmentDetails!!.chunkUpdates[i]
//                info.downloaded = s.downloaded
//                info.start = s.startOffset
//                info.length = s.length
//            }
//        }
//    }
//}
